### PR TITLE
gh-86493: Modernize modules initialization code

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
Use PyModule_Add() or PyModule_AddObjectRef() instead of soft deprecated PyModule_AddObject().

PyModule_AddObject() is only left in two modules: `_testcapi` and `_testbuffer`. But that code ignores any failures for now. They need more significant rewriting.

* Issue: gh-86493

----
:books: Documentation preview :books:: https://cpython-previews--106858.org.readthedocs.build/

### CODE REVIEW

File: Doc/extending/newtypes_tutorial.rst [hunk 1]
```diff
@@ -180,9 +180,7 @@ This initializes the :class:`Custom` type, filling in a number of members
 to the appropriate default values, including :attr:`ob_type` that we initially
 set to ``NULL``. ::
 
-   Py_INCREF(&CustomType);
-   if (PyModule_AddObject(m, "Custom", (PyObject *) &CustomType) < 0) {
-       Py_DECREF(&CustomType);
+   if (PyModule_AddObjectRef(m, "Custom", (PyObject *) &CustomType) < 0) {
```
Comment (vstinner) [inline Doc/extending/newtypes_tutorial.rst:183]: PyModule_AddType() may be more appropriate. Same below.

File: Modules/_tkinter.c [hunk 2]
```diff
@@ -3207,17 +3207,11 @@ PyInit__tkinter(void)
     if (m == NULL)
         return NULL;
```
Comment (vstinner) [inline Modules/_tkinter.c:3209]: GHA compiler warning: unused variable ‘o’ [-Wunused-variable]

File: Modules/_tkinter.c [hunk 3]
```diff
@@ -3207,17 +3207,11 @@ PyInit__tkinter(void)
...
 
-    o = PyErr_NewException("_tkinter.TclError", NULL, NULL);
-    if (o == NULL) {
+    Tkinter_TclError = PyErr_NewException("_tkinter.TclError", NULL, NULL);
+    if (PyModule_AddObjectRef(m, "TclError", Tkinter_TclError)) {
         Py_DECREF(m);
```
Comment (vstinner) [inline Modules/_tkinter.c:3212]: Note for later: i prefer "exec" functions which returns -1 on error, and the caller is responsible to manage the module refcount.

File: Modules/binascii.c [hunk 4]
```diff
@@ -1261,24 +1261,12 @@ binascii_exec(PyObject *module) {
...
 
     state->Error = PyErr_NewException("binascii.Error", PyExc_ValueError, NULL);
-    if (state->Error == NULL) {
-        return -1;
-    }
-    Py_INCREF(state->Error);
-    result = PyModule_AddObject(module, "Error", state->Error);
-    if (result == -1) {
-        Py_DECREF(state->Error);
+    if (PyModule_AddObjectRef(module, "Error", state->Error) < 0) {
```
Comment (vstinner) [inline Modules/binascii.c:1264]: GHA compiler warning: unused variable ‘result’ [-Wunused-variable]

File: Modules/xxlimited_35.c [hunk 5]
```diff
@@ -247,43 +245,24 @@ xx_modexec(PyObject *m)
...
     }
 
     /* Add Xxo */
     Xxo_Type = PyType_FromSpec(&Xxo_Type_spec);
-    if (Xxo_Type == NULL) {
-        return -1;
-    }
-    if (PyModule_AddObject(m, "Xxo", Xxo_Type) < 0) {
-        Py_DECREF(Xxo_Type);
+    if (PyModule_Add(m, "Xxo", Xxo_Type) < 0) {
```
Comment (vstinner) [inline Modules/xxlimited_35.c]: It's not part of the limited C API 3.5: GHA compiler warning: implicit declaration of function ‘PyModule_Add’; did you mean ‘PyModule_New’? [-Werror=implicit-function-declaration]

File: Modules/zlibmodule.c [hunk 6]
```diff
@@ -2030,17 +2030,11 @@ zlib_exec(PyObject *mod)
...
-
-    if (PyModule_AddObject(mod, "error", Py_NewRef(state->ZlibError)) < 0) {
-        Py_DECREF(state->ZlibError);
-        return -1;
-    }
-    if (PyModule_AddObject(mod, "_ZlibDecompressor",
-                           Py_NewRef(state->ZlibDecompressorType)) < 0) {
-        Py_DECREF(state->ZlibDecompressorType);
+    if (PyModule_AddObjectRef(mod, "_ZlibDecompressor",
+                           state->ZlibDecompressorType) < 0) {
```
Comment (vstinner) [inline Modules/zlibmodule.c]: GHA warning: passing argument 3 of ‘PyModule_AddObjectRef’ from incompatible pointer type [-Wincompatible-pointer-types]

File: Modules/zlibmodule.c [hunk 7]
```diff
@@ -2082,26 +2076,12 @@ zlib_exec(PyObject *mod)
 #ifdef Z_TREES // 1.2.3.4, only for inflate
     ZLIB_ADD_INT_MACRO(Z_TREES);
 #endif
-    PyObject *ver = PyUnicode_FromString(ZLIB_VERSION);
-    if (ver == NULL) {
+    if (PyModule_Add(mod, "ZLIB_VERSION", PyUnicode_FromString(ZLIB_VERSION)) < 0) {
```
Comment (vstinner) [inline Modules/zlibmodule.c]: Maybe put PyUnicode_From... on a second line for readability, same below.

**Review by vstinner — APPROVED** [review]
LGTM, but a few compiler warnings should be fixed first: see GHA job results, especially the Ubuntu job which logged many warnings.

File: Doc/extending/newtypes_tutorial.rst (hunk 1 above)
Comment (serhiy-storchaka) [reply Doc/extending/newtypes_tutorial.rst:183]: `PyModule_AddType()` also calls `PyType_Ready()`. This makes the above few lines unnecessary.

I think that this tutorial should be rewritten with examples of appropriate use of `PyModule_AddObjectRef()`, `PyModule_Add()` and `PyModule_AddType()`. Maybe even `PyModule_AddInt()` and `PyModule_AddFunction()`. But this is a different large issue. For now, I only want to remove deprecated `PyModule_AddObject()` from examples.

Unlike to `PyModule_AddObjectRef()`, `PyModule_AddType()` cannot be called with NULL, so it is not possible to replace all semantically appropriate uses of `PyModule_AddObjectRef()` with `PyModule_AddType()`. Maybe after making the latter working with NULL...

File: Modules/_lzmamodule.c [hunk 8]
```diff
@@ -1498,15 +1498,7 @@ _lzma__decode_filter_properties_impl(PyObject *module, lzma_vli filter_id,
...
 {
-    PyObject *o = PyLong_FromLongLong(value);
-    if (o == NULL) {
-        return -1;
-    }
-    if (PyModule_AddObject(m, name, o) == 0) {
-        return 0;
-    }
-    Py_DECREF(o);
-    return -1;
```
Comment (serhiy-storchaka) [inline Modules/_lzmamodule.c:1509]: It is very confusing code with inverse logic. I thought there was a bug.

File: Modules/_testcapi/structmember.c [hunk 9]
```diff
@@ -193,7 +193,7 @@ _PyTestCapi_Init_Structmember(PyObject *m)
     if (res < 0) {
         return -1;
     }
-    res = PyModule_AddObject(
+    res = PyModule_AddObjectRef(
```
Comment (serhiy-storchaka) [inline Modules/_testcapi/structmember.c]: I think there was a bug.

File: Modules/xxlimited_35.c [hunk 10]
```diff
@@ -247,43 +245,24 @@ xx_modexec(PyObject *m)
...
     if (ErrorObject == NULL) {
         ErrorObject = PyErr_NewException("xxlimited_35.error", NULL, NULL);
-        if (ErrorObject == NULL) {
-            return -1;
-        }
     }
-    Py_INCREF(ErrorObject);
-    if (PyModule_AddObject(m, "error", ErrorObject) < 0) {
-        Py_DECREF(ErrorObject);
+    if (PyModule_AddObjectRef(m, "error", ErrorObject) < 0) {
```
Comment (serhiy-storchaka) [inline Modules/xxlimited_35.c]: It is interesting, that the compiler complains about `PyModule_Add`, but not about `PyModule_AddObjectRef`. There is a bug in `PyModule_AddObjectRef` declaration.

**Review by corona10 — APPROVED** [review]
lgtm too

File: Doc/extending/newtypes_tutorial.rst (hunk 1 above)
Comment (vstinner) [reply Doc/extending/newtypes_tutorial.rst:183]: Ok, it makes sense.

**Review by vstinner — APPROVED** [review]
LGTM. I didn't check individual refcount, I rely on Refleaks buildbots for that and other reviewers 😁 Overall the change LGTM and makes the code shorter and more regular. The old code was really hard to understand with INCREF/DECREF dance.

**Comment by vstinner** [comment]
That's a big change, thanks.

**Pull request merged**